index.ts

```diff
@@ -0,0 +1,141 @@
+import { FastifyPluginAsync } from "fastify";
+import { Actor, Item, Member, PostHookHandlerType, PreHookHandlerType, UnknownExtra } from "graasp";
+
+interface GraaspS3FileItemOptions {
+  type: string;
+  sizePath: string;
+}
+
+type MemberExtra = {
+  storage: {
+    total: number;
+  };
+};
+
+const DEFAULT_MAX_STORAGE = 1024 * 1024 * 1024 * 15; // 15GB;
+
+class StorageExceeded extends Error {
+  message = "The allowed storage is full";
+  name = "StorageExceededError";
+  origin = "plugin";
+  code = 507;
+}
+
+const plugin: FastifyPluginAsync<GraaspS3FileItemOptions> = async (fastify, options) => {
+  const {
+    items: { taskManager },
+    members: { taskManager: mTaskManager, dbService: mDbService },
+    taskRunner: runner,
+    log,
+  } = fastify;
+  const { type: itemType, sizePath } = options;
+
+  if (!itemType || !sizePath) {
+    throw new Error("graasp-file-upload-limiter: missing plugin options");
+  }
+
+  const getMemberStorageFromExtra = (extra: MemberExtra) => extra?.storage?.total || 0;
+
+  const getFileSize = (item: Partial<Item<UnknownExtra>>): number => {
+    const properties = sizePath.split(".");
+    const sizeField = properties.reduce((prev, curr) => prev?.[curr], item);
+
+    if (Number.isInteger(sizeField)) {
+      return sizeField;
+    }
+
+    throw new Error("sizePath does not lead to a size number");
+  };
+
+  // increase the actor storage by the given item size
+  const increaseStorage: PostHookHandlerType<Item<UnknownExtra>> = async (item, actor) => {
+    const { type } = item;
+
+    // enabled only on given item type
+    if (type !== itemType) return;
+
+    const size = getFileSize(item);
+    // the file size is not available
+    if (size <= 0) return;
+
+    const member = actor as Member<MemberExtra>;
+    let currentStorage = getMemberStorageFromExtra(member.extra);
+
+    const task = mTaskManager.createUpdateTask(actor, actor.id, {
```

Comment:
Humm .... I didn't foresee, or thought about, the way of doing this, but this way _no me gusta_: tasks running inception. Let's discuss it.